### ISSUE TITLE
holes in read_country() with no arguments

### ISSUE THREAD

**Issue by pedro-andrade-inpe** [ipeaGIT/geobr#41]
read_country() with no arguments returns the geometry of Brazil in 2010 with several holes. The data for 2014 and 2015 also have some holes. Perhaps removing the holes manually after computing the union solves the problem as shown in the code below.

(I saw in another issue the discussion that st_union is very slow. unionSpatialPolygons from maptools is much faster)

```R
require(geobr)
require(dplyr)
require(sp)
require(sf)
require(maptools)

sp_states <- read_state(year=2010, code_state = "all") %>% as("Spatial")

result <- unionSpatialPolygons(sp_states, rep(TRUE, 27))

outerRings = Filter(function(f){f@ringDir==1},result@polygons[[1]]@Polygons)
outerBounds = SpatialPolygons(list(Polygons(outerRings,ID=1)))
plot(outerBounds)

m <- st_as_sf(outerBounds)

write_sf(m, "brazil.shp")
```

**Comment by rafapereirabr** [comment]
Thanks for the suggestions, Pedro! That problem is really bugging me as well. It will take me a few days to implement this solution, though,  because I need to generate the data sets for every year again incorporating your suggestion. Only after that I can update the files in the package directory so the package returns the data properly cleaned. I'll be away from my office in the next few days, but I should be able to incorporate your suggestions by the end of the month.

**Comment by rafapereirabr** [comment]
Pedro, I've used some of your code to update the scrip `./prep_data/prep_country.R`, which we use to generate the data sets for the `read_country()` function. Your solution worked well for the 2010 data. However, it didn't work for 2000 (I haven't tested for other years before 2005).

In the year 2000, for example, I get this error when I try to conver the `sf` object to `sp`. the


> Error in sp::Polygons(lapply(seq_along(x), function(i) sp::Polygon(x[[i]],  : 
>   length(srl) > 0 is not TRUE

```
    temp_sf <- read_state(year=2000, code_state = "all")


  # b) make sure we have valid geometries
    temp_sf <- lwgeom::st_make_valid(temp_sf)
    temp_sf <- temp_sf %>% st_buffer(0)

    # convert to sp
      temp_sp <- sf::as_Spatial(temp_sf)
      temp_sp <- rgeos::gBuffer(temp_sp, byid=TRUE, width=0)


  # c) dissolve borders to create country file
    result <- maptools::unionSpatialPolygons(temp_sp, rep(TRUE, nrow(temp_sp@data))) # dissolve


  # d) get rid of holes
    outerRings = Filter(function(f){f@ringDir==1},result@polygons[[1]]@Polygons)
    outerBounds = SpatialPolygons(list(Polygons(outerRings,ID=1)))

plot(outerBounds 
```

I will have more time to look into latter (probably next month) , but let me know if you have other suggestions

**Comment by pedro-andrade-inpe** [comment]
There are empty polygons that need to be removed in order to be properly converted to sp. I implemented the following code to remove them (not a beautiful solution but it works):

```R

sp_states_2000 <- read_state(year=2000, code_state = "all")
temp_sf <- lwgeom::st_make_valid(sp_states_2000)
temp_sf <- temp_sf %>% st_buffer(0)

sp_states1 <- temp_sf %>% st_cast("MULTIPOLYGON")

# create attribute with the number of points each polygon has
points_in_each_polygon = sapply(1:dim(sp_states1)[1], function(i)
  length(st_coordinates(sp_states1$geometry[i])))

sp_states1$points_in_each_polygon <- points_in_each_polygon


mypols <- sp_states1 %>% filter(points_in_each_polygon > 0)

sp_states_2000a <- mypols %>% as("Spatial")

```

**Comment by rafapereirabr** [comment]
Great. I’ll have a look at this when I’m back in my office by the end of the month. Thanks again 

Rafael H.M. Pereira

**Comment by rafapereirabr** [comment]
Fixed. Thanks again for the support, Pedro!

**Comment by tcs-git** [comment]
I believe I have a similar problem to aggregate the municipalities. Run for hours until it crashes.
Do you have any correction suggestions?
Thank you!

`
mun <- geobr::read_municipality(year=2020)

mun_agg <- mun %>% 
  dplyr::as_tibble() %>% 
  sf::st_as_sf() %>% 
  dplyr::group_by(abbrev_state) %>% 
  dplyr::summarise()
`

**Comment by rafapereirabr** [comment]
Hi @tcs-git . Two quick comments here:
1. If you want to get the geometries of states, it would be better to download them directly with the `read_state()` function instead of doing a summarise operation.

3. Nonetheless, if you really need to do a summarise operation on municipal geometries, I would strongly suggest you download them using the parameter `simplified = FALSE`.
``` 

mun <- geobr::read_municipality(year=2020, simplified = FALSE)
```

**Comment by tcs-git** [comment]
1. The grouping for states was just an example. I actually need to run a custom grouping of municipalities.

2. With the parameter Simplified = F the cluster only ran if I use sf::sf_use_s2(F), otherwise I have a problem with spherical geography on the map. Since my goal is just to create a viz, I don't think this is a problem.

This was the code that worked for me, after your help:

```
mun <- geobr::read_municipality(year=2020, simplified = F)

sf::sf_use_s2(F)

mun_agg <- mun %>%
dplyr::as_tibble() %>%
sf::st_as_sf() %>%
dplyr::group_by(abbrev_state) %>%
dplyr::summarise()
```

Thanks
